packages/aggregator/src/analytics/aggregateAnalytics/aggregations/entityAggregations.js

```diff
@@ -14,20 +14,25 @@ import { checkValueSatisfiesCondition } from '@tupaia/utils';
  * @param {Function} keyMapper
  */
 const valuePerOrgGroup = (analytics, aggregationConfig, valueMapper, keyMapper) => {
-  const { orgUnitMap = {}, condition } = aggregationConfig;
+  const { orgUnitMap = {}, condition, periodType } = aggregationConfig;
```

Comment:
Added periodType to allow converting to desired periods when doing entityAggregation

Reply:
Can we support this by using a series of aggregations instead? It seems unrelated to the entity aggregation process.

Reply:
True it does seem unrelated to entity aggregation! So there are 2 aggregations performed in these dashboards (listed here: https://github.com/beyondessential/tupaia/pull/1942#discussion_r563420278):
- FINAL_EACH_QUARTER
- COUNT_PER_PERIOD_PER_ORG_GROUP (entity aggregation)

I just need the `period` of the analytics to be in `quarter` format before entityAggregation is performed, so that' why I tried to convert the periods by adding `periodType` here.. Perhaps I can do that in the `FINAL_EACH_QUARTER` aggregation instead?

Reply:
@kael89 you had some thoughts on converting periods to their correct formats during aggregation right? I believe doing it in `FINAL_EACH_QUARTER` would fit with what you were thinking.

Reply:
Definitely, it seems that most existing aggregations are not concerned with changing period formats, although their name would imply so. It also a blocker for using most `FINAL_EACH...` aggregations with `indicators`.

Card: https://github.com/beyondessential/tupaia-backlog/issues/2150#issue-788619473 - 4th point ("Make periods returned...")

---

My only other comment is that `FINAL_EACH_WEEK` works correctly, since it **has** to convert periods to the special week format (eg 2021W3), before comparing them. This is not a requirement for `FINAL_EACH_DAY`, `FINAL_EACH_MONTH`, `FINAL_EACH_YEAR`, which are just an extension one another (2020 => 202001 => 20200101), and can use a single string comparison.

Based on the above observation, I would expect that `FINAL_EACH_QUARTER` also **has** to convert periods to its custom format (2021Q1), but I may be wrong.

Reply:
> most existing aggregations are not concerned with changing period formats, although their name would imply so

I did have the same thought... I'm very tempted to make this global change for `FINAL_EACH...` aggregations. However, I'm also afraid of the amount of regression testing... (given that this card already requires a lot of regression test). In my opinion, Having a config seems like the best option for now?

Reply:
I reckon the fact that this card already requires regression test is an argument _for_ changing it now. I'm not sure how likely regressions are though, being unfamiliar with the data building code these days - do you know whether there are many places that assume the format of a period? @kael89, do you?

Reply:
Possibly [getPreferredPeriod](https://github.com/beyondessential/tupaia/blob/dev/packages/aggregator/src/analytics/aggregateAnalytics/aggregations/utils/getPreferredPeriod.js)... I believe that there are a few reports where is supplied by both `mSpply` and `meditrak`. The one set of data uses a daily period, and the other a monthly. We prefer one over the other for this specific report.

If those reports use `FINAL_EACH` aggregations, and those aggregations change the period formats, the above comparison may no longer be possible

If this functionality is indeed desirable, I suggest adding an optional config option to `FINAL_EACH`, eg 
```js
{ useOriginalPeriodType: true }
// or
{ retainPeriodType: true }
```

Reply:
Sorry for my late reply, I was interrupted with some other tasks :(. 

So I've had a search on the reports and found at least 1 report that uses both `mSupply` and `meditrak` data element codes and also use `FINAL_EACH_...` (id: `UNFPA_Delivery_Services_Stock`). Possibly in the future we can have more? So I'm keen to follow Kostas' suggestion

Reply:
Thanks for the investigation @chris-bes , did you also confirm that this report's functionality depends on each source using different period types in analytics?

Reply:
That's a good point... Now that I've checked, the data coming from `mSupply` does have daily period rather than monthly period that we expected. Could it be that the monthly period doesn't apply to all mSupply data elements?
![msupply](https://user-images.githubusercontent.com/63621318/106423346-61a49900-64b4-11eb-9912-a7e30e829246.PNG)

Reply:
Just had more investigations on this. Perhaps the reports that @kael89 was talking about are the ones that use `Medicines Availability Aggregation`. They use a preaggregated element, so that's why I failed to search for it in the first place. To pre aggregate the data elements, they use `FINAL_EACH_MONTH_PREFER_DAILY_PERIOD` aggregation. So I think this is the case...

Reply:
That must be it @chris-bes !!

Reply:
It seems quite isolated - the meds availability preaggregation is the only place I know of where we use the difference in period types as information, and it seems it's built in to the aggregation type so will be able to happen before the conversion to uniform period type. Does this mean we can go with consistently converting to the expected period type, rather than using a config option? I'm nervous of the proliferation of config options!

Reply:
I agree we should avoid adding an extra option only for a specific use case. Even if we leverage the difference of period types in the input, we can still sanitize all periods in the output according to the `FINAL_EACH....` period type.

The only problem would be if someone runs a `FINAL_EACH_MONTH` aggregation and then a `FINAL_EACH_MONTH_PREFER_DAILY_PERIOD` , but it doesn't seem to be the case in this report.

Reply:
Thanks guys for your valuable inputs! I can have a try at making that global change (without using config to convert the period types) and will do a smoke test on the reports!